### DISCUSSION TITLE
iOS Error: No APNS token specified before fetching FCM Token

### DISCUSSION THREAD

**Discussion by gz1992** [invertase/react-native-firebase#7329] [Q&A] [answered]
Hey guys, I am creating a new app using version 0.70.6 and @react-native-firebase/messaging ^17.3.1

But I have an old react native app version 0.62.3 using @react-native-firebase/messaging ^15.6.0 on App Store. 

I am facing some problems regarding FCMToken, especially this one: No APNS token specified before fetching FCM Token.
When I try to login in my app. For Android, everything went well, but iOS didn't.

Does anybody know how can I test on my computer without getting this "No APNS token specified before fetching FCM Token" problem?

OBS: My Xcode is updated properly, because on my TestFlight app, I can login and also send push notifications

**Comment by gz1992** [comment]
I think I found the problem. I changed the way my APNS token is used in my iOS, via firebase. And I need to force the update of all FCMToken that are used in the previous way. Does somebody know how to do it??

**Comment by gz1992** [comment, marked as answer]
Found it... As FCMToken is related to the user device, when I changed my certificate on firebase, some users that couldn't update their token couldn't access my platform. It is important to mention that the users had permission but couldn't get token
```
messaging().hasPermission().then(permission => {})
```

As a result, I needed to force the registration of their device by:
```
await messaging().registerDeviceForRemoteMessages();
return messaging()
    .getToken()
    .then(token => token)
```

**Reply by hadanghuy** [reply]
Bạn có phần sửa cụ thể ở chỗ nào không?

**Reply by tuannha01** [reply]
@gz1992 Can you tell hime where to updated 😄

**Reply by gz1992** [reply]
hey @tuannha01 and @hadanghuy , in my case, we migrated and updated the firebase lib, and we had a problem with the fcmtoken, the cellphone was already registered... so instead **ASKING AND IDENTIFING THE PERMISSION**
```
messaging().hasPermission().then(permission => {})
```

i  needed to force the **device registry again**, through this code:
```
await messaging().registerDeviceForRemoteMessages();
return messaging()
    .getToken()
    .then(token => token)
```

**Comment by mhassanmalik** [comment]
My issue was resolved by adding onRegister() inside PushNotification.configure(). I was not setting the retrieved APNS token in messaging() before asking for the FCM token.

```
onRegister(token) {
  if (Platform.OS === 'ios') {
    messaging().setAPNSToken(token.token);
  }
}
```

**Comment by Rakhilpr96** [comment]
**Ensure FirebaseAppDelegateProxyEnabled is not disabled**

In `Info.plist`, check:

```xml
<key>FirebaseAppDelegateProxyEnabled</key>
<true/>
```

If you set it to false, you must manually forward APNs token like this in `AppDelegate.mm`:
```objc
- (void)application:(UIApplication *)application didRegisterForRemoteNotificationsWithDeviceToken:(NSData *)deviceToken {
  [FIRMessaging messaging].APNSToken = deviceToken;
}
```

**Comment by tonsnoei** [comment]
If you forget to set the [background modes capabilities](https://rnfirebase.io/messaging/usage/ios-setup) in your iOS project, this problem will occur too.

**Comment by salil1607** [comment]
If you are using new architeture then update AppDelegate.swift

```swift
    UNUserNotificationCenter.current().delegate = self
    UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, _ in
      if granted {
        DispatchQueue.main.async {
          application.registerForRemoteNotifications()
        }
      }
    }

    application.registerForRemoteNotifications()
    Messaging.messaging().delegate = self
    
    
  func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
    Messaging.messaging().apnsToken = deviceToken
  }

  func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
    NSLog("Remote notification registration failed: \(error.localizedDescription)")
  }

  func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?) {
    // Token provided to JavaScript via RNFirebase bridge when requested.
  }
}
```

**Comment by devsynschismo** [comment]
I tested in IOS simulator then got this error
Error getting FCM token: Error: [messaging/unknown]。No APNS token specified before fetching FCM Token

I just turn off the simulator then turn it on again, the error dissapear :)

**Comment by Fabiano-Bragaaa** [comment]
In my case, this error was caused by corrupted iOS Simulator state. No code changes were needed — the existing setup was already
  correct.                                                                                                                             
                                                                                                                                       
  Fix: force quit the simulator → clean Xcode build folder (Product → Clean Build Folder) → rebuild and run directly through Xcode with
   Metro running separately.                                                                                                           
                  
  My setup (modular API):                                                                                                              
  import {        
    getMessaging,                                                                                                                      
    requestPermission,
    getToken,         
    AuthorizationStatus,
  } from '@react-native-firebase/messaging';
                                            
  async function requestNotificationPermission() {                                                                                     
    const messaging = getMessaging();             
                                                                                                                                       
    if (Platform.OS === 'ios') {
      const authStatus = await requestPermission(messaging);
      const enabled =                                       
        authStatus === AuthorizationStatus.AUTHORIZED ||
        authStatus === AuthorizationStatus.PROVISIONAL;                                                                                
                                                       
      if (!enabled) return;                                                                                                            
    } else {               
      await PermissionsAndroid.request(
        PermissionsAndroid.PERMISSIONS.POST_NOTIFICATIONS,
      );                                                  
    }                                                                                                                                  
     
    const token = await getToken(messaging);                                                                                           
    console.log('FCM Token:', token);       
  }                                  
   
  Versions: React Native 0.81.5, @react-native-firebase/app 23.8.8, @react-native-firebase/messaging 23.8.8
